### PR TITLE
Fix segfault in replication_pad{1d,2d}_backward on channel mismatch

### LINKED ISSUE #142834: Segmentation fault in `replication_pad2d_backward`
### 🐛 Describe the bug

Under specific inputs, `replication_pad2d_backward` triggered a crash.
```python
import torch

grad_output = torch.full((2, 0, 6, 8,), 1, dtype=torch.float)
self = torch.full((2, 2, 4, 4,), 1, dtype=torch.float)
padding = [2, 2, 1, 1]
torch.ops.aten.replication_pad2d_backward(grad_output, self, padding)
```
Output
```
Segmentation fault (core dumped)
```

### Versions

PyTorch version: 2.5.0a0+git32f585d
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: Ubuntu 22.04.4 LTS (x86_64)
GCC version: (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0
Clang version: Could not collect
CMake version: version 3.22.1
Libc version: glibc-2.35

Python version: 3.13.0 | packaged by Anaconda, Inc. | (main, Oct 7 2024, 21:29:38) [GCC 11.2.0] (64-bit runtime)
Python platform: Linux-5.15.0-124-generic-x86_64-with-glibc2.35
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture: x86_64
CPU op-mode(s): 32-bit, 64-bit
Address sizes: 46 bits physical, 48 bits virtual
Byte Order: Little Endian
CPU(s): 80
On-line CPU(s) list: 0-79
Vendor ID: GenuineIntel
Model name: Intel(R) Xeon(R) Gold 5218R CPU @ 2.10GHz
CPU family: 6
Model: 85
Thread(s) per core: 2
Core(s) per socket: 20
Socket(s): 2
Stepping: 7
CPU max MHz: 4000.0000
CPU min MHz: 800.0000
BogoMIPS: 4200.00
Flags: fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf pni pclmulqdq dtes64 monitor ds_cpl vmx smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid dca sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault epb cat_l3 cdp_l3 invpcid_single intel_ppin ssbd mba ibrs ibpb stibp ibrs_enhanced tpr_shadow vnmi flexpriority ept vpid ept_ad fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid cqm mpx rdt_a avx512f avx512dq rdseed adx smap clflushopt clwb intel_pt avx512cd avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local dtherm ida arat pln pts pku ospke avx512_vnni md_clear flush_l1d arch_capabilities
Virtualization: VT-x
L1d cache: 1.3 MiB (40 instances)
L1i cache: 1.3 MiB (40 instances)
L2 cache: 40 MiB (40 instances)
L3 cache: 55 MiB (2 instances)
NUMA node(s): 2
NUMA node0 CPU(s): 0,2,4,6,8,10,12,14,16,18,20,22,24,26,28,30,32,34,36,38,40,42,44,46,48,50,52,54,56,58,60,62,64,66,68,70,72,74,76,78
NUMA node1 CPU(s): 1,3,5,7,9,11,13,15,17,19,21,23,25,27,29,31,33,35,37,39,41,43,45,47,49,51,53,55,57,59,61,63,65,67,69,71,73,75,77,79
Vulnerability Gather data sampling: Mitigation; Microcode
Vulnerability Itlb multihit: KVM: Mitigation: VMX disabled
Vulnerability L1tf: Not affected
Vulnerability Mds: Not affected
Vulnerability Meltdown: Not affected
Vulnerability Mmio stale data: Mitigation; Clear CPU buffers; SMT vulnerable
Vulnerability Reg file data sampling: Not affected
Vulnerability Retbleed: Mitigation; Enhanced IBRS
Vulnerability Spec rstack overflow: Not affected
Vulnerability Spec store bypass: Mitigation; Speculative Store Bypass disabled via prctl and seccomp
Vulnerability Spectre v1: Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2: Mitigation; Enhanced / Automatic IBRS; IBPB conditional; RSB filling; PBRSB-eIBRS SW sequence; BHI SW loop, KVM SW loop
Vulnerability Srbds: Not affected
Vulnerability Tsx async abort: Mitigation; TSX disabled

Versions of relevant libraries:
[pip3] numpy==2.1.3
[pip3] torch==2.5.0a0+git32f585d
[conda] numpy 2.1.3 pypi_0 pypi
[conda] torch 2.5.0a0+git32f585d pypi_0 pypi

### PR DESCRIPTION
**Pull request by Ansh-Karnwal**
Fixes #142834.

`replication_pad2d_backward` (and the 1d variant) segfaulted when `gradOutput`'s channel dimension did not match the input's. The backward shape validation checked the spatial dims (width/height) but never the channel dim, so a `gradOutput` with a mismatched channel count (e.g. 0 channels) passed the checks, a non-empty `gradInput` was allocated from the input's shape, and the kernel then read `gradOutput` out of bounds -> SIGSEGV.

The 3d backward already validates the channel dim; this mirrors that check into the 1d backward (in its `TORCH_META_FUNC`, so it covers all backends) and the 2d backward (both the CPU template and the CUDA template). All three now raise a clear `gradOutput channel unexpected` error instead of crashing. Legitimate empty and no-batch inputs are unaffected because their channel dims still match.

## Test plan

The reporter's reproducer now raises a clear error instead of crashing, on both CPU and CUDA:

```python
torch.ops.aten.replication_pad2d_backward(torch.full((2, 0, 6, 8), 1.), torch.full((2, 2, 4, 4), 1.), [2, 2, 1, 1])
```

Added a device-generic regression test (`test_ReplicationPad_backward_channel_mismatch`, covering 1d and 2d) and ran the full ReplicationPad suite:

```
python test/test_nn.py -k ReplicationPad -v
```

All 18 ReplicationPad tests pass on CPU and CUDA, including the existing empty-input and no-batch-dim tests (confirming valid inputs are not rejected).

### CODE REVIEW

File: aten/src/ATen/native/ReplicationPadding.cpp [hunk 1]
```diff
@@ -187,20 +193,26 @@ void replication_pad2d_backward_out_cpu_template(
...
   }
 
   /* sizes */
+  int64_t ichannel = input.size(dimc);
   int64_t iheight = input.size(dimh);
   int64_t iwidth = input.size(dimw);
   int64_t oheight = iheight + pad_t + pad_b;
   int64_t owidth  = iwidth + pad_l + pad_r;
 
+  TORCH_CHECK(ichannel == gradOutput.size(dimc),
```
Comment (Skylion007) [inline aten/src/ATen/native/ReplicationPadding.cpp:213]: These should all be ValueErrors, but that is a different PR. To confirm, it segfaulted even when the channel size was 1 (Ie for broadcasting?)

Comment (Ansh-Karnwal) [reply aten/src/ATen/native/ReplicationPadding.cpp:213]: Thanks for taking a look!

It's not a broadcasting case: `replication_pad` preserves the channel count in the forward (a C=1 input gives a C=1 output), so a valid `grad_output` has to match the input's channels exactly. And yes, it segfaulted for any mismatch, not just channel 0: a C=1 `grad_output` against a C=2 input hits the same out-of-bounds read (the kernel loops over the input's channels and indexes into `grad_output`). That case now raises `gradOutput channel unexpected. Expected: 2, Got: 1`, while a genuine C=1 input still works.

And agreed on `ValueError` - happy to do that as a follow-up.

**Review by Skylion007 — APPROVED** [review]

**Comment by Ansh-Karnwal** [comment]
Addressed the Copilot feedback: scoped the regression test with `@skipMPS` (MPS backward goes through a separate kernel, `mps::pad_out_template`, that does not validate the channel dim), and fixed the missing comma in the 1d error message for consistency with the 2d/3d checks.

@Skylion007 could you approve the workflows / re-run CI on the new commit when you get a chance? Thanks!

**Comment by Ansh-Karnwal** [comment]
looks like the 3 failed and 3 cancelled jobs are all CI infrastructure issues (container start / credentials failures, and Kubernetes `CreateContainerError` no tests actually ran), not related to this change; the real test shards all passed. Rebasing to re-run on healthy runners.

@pytorchbot rebase

**Comment by Ansh-Karnwal** [comment]
@Skylion007 the bot won't let me rebase (first-time contributor), and the 3 failed + 3 cancelled jobs are all CI-infra failures (container start / credentials / `CreateContainerError` -- no tests actually ran); all the real test shards passed. Could you re-run the failed jobs or merge through them when you get a chance? Thanks! It's my first time so hope its not a bother.
